packages/dd-trace/src/appsec/rasp/sql_injection.js

```diff
@@ -0,0 +1,64 @@
+'use strict'
+
+const { pgQueryStart, pgPoolQueryStart, pgPoolQueryFinish } = require('../channels')
+const { storage } = require('../../../../datadog-core')
+const addresses = require('../addresses')
+const waf = require('../waf')
+const { RULE_TYPES, handleResult } = require('./utils')
+
+let config
+
+function enable (_config) {
+  config = _config
+
+  pgQueryStart.subscribe(analyzePgSqlInjection)
+  pgPoolQueryStart.subscribe(analyzePgSqlInjectionInPool)
+  pgPoolQueryFinish.subscribe(pgPoolFinish)
+}
+
+function disable () {
+  if (pgQueryStart.hasSubscribers) pgQueryStart.unsubscribe(analyzePgSqlInjection)
+  if (pgPoolQueryStart.hasSubscribers) pgPoolQueryStart.subscribe(analyzePgSqlInjectionInPool)
+  if (pgPoolQueryFinish.hasSubscribers) pgPoolQueryFinish.subscribe(pgPoolFinish)
+}
+
+function analyzePgSqlInjection (ctx) {
+  const query = ctx.query?.text
+  if (!query) return
+
+  const store = storage.getStore()
+  if (!store) return
+
+  const { raspSqlAnalyzed, req } = store
+
+  if (!req || raspSqlAnalyzed) return
+
+  const persistent = {
+    [addresses.DB_STATEMENT]: query,
+    [addresses.DB_SYSTEM]: 'postgresql' // TODO: Extract to constant
+  }
+
+  const result = waf.run({ persistent }, req, RULE_TYPES.SQL_INJECTION)
+
+  const res = store?.res
+  handleResult(result, req, res, ctx.abortController, config)
+}
+
+function analyzePgSqlInjectionInPool (ctx) {
+  const parentStore = storage.getStore()
+  if (!parentStore) return
+
+  analyzePgSqlInjection(ctx)
+
+  storage.enterWith({ ...parentStore, raspSqlAnalyzed: true, raspSqlParentStore: parentStore })
```

Comment:
If finish is not executed, it is a instrumentation bug, and it will be a problem also for the tracer spans, because pg span is not going to be closed.

> what if it's not called ?

I guess that worst case, we are not going to analyze more sql calls in the same request, because we are not going t orestore `raspSqlAnalyzed` property. 

> Are you trying to make sure the query is not analyzed twice ?
Yes, don't want to analyze twice the same query, in pg.Pool and pg.Query, because pg-pool calls internally to pg query.

> in that case instead of passing a weird store around you could have a weakSet ?

I'll try to do something like this... but I don't think that I could have a key object that would be unique and same instance in pool and query. I'll try to find something.

Maybe having `req` as key of the weakMap and `Set<String>` with queries as value...

Reply:
not sure the query is unique enough.
Another thing to take into consideration, enterWith() is not part of the tc39 async context spec, so we should probably avoid to rely on it in new code, as it probably won't be here forever